File: algorithms/trivial_structs/queue_deque.py
# Queue / Deque:
# Time Complexity:
# Enqueue:      O(1)
# Dequeue:      O(1)
# Peek:         O(1)
#
# append():     O(1)
# appendleft(): O(1)
# pop():        O(1)
# popleft():    O(1)
# peek():       O(1)
#
# Space Complexity:
# Memory:       O(n)

class Deque:
    def __init__(self, block_size=4):
        self.block_size = block_size
        self.blocks = [[]]

    def append(self, value):
        if len(self.blocks[-1]) == self.block_size:
            self.blocks.append([])

        self.blocks[-1].append(value)

    def appendleft(self, value):
        if len(self.blocks[0]) == self.block_size:
            self.blocks.insert(0, [])

        self.blocks[0].insert(0, value)

    def pop(self):
        if not self.blocks:
            raise IndexError

        val = self.blocks[-1].pop()

        if not self.blocks[-1]:
            self.blocks.pop()

        return val

    def popleft(self):
        if not self.blocks:
            raise IndexError

        val = self.blocks[0].pop(0)

        if not self.blocks[0]:
            self.blocks.pop(0)

        return val

    def peek_left(self):
        return self.blocks[0][0]

    def peek_right(self):
        return self.blocks[-1][-1]

class Queue:
    def __init__(self):
        self.data = Deque()

    def enqueue(self, value):
        self.data.append(value)

    def dequeue(self):
        return self.data.popleft()

    def peak(self):
        return self.data.peek_left()

# Nothing out of the ordinary—it’s a standard list with FIFO by default,
# although you can do whatever you like with a deque via collections.
# A VERY HANDY STRUCTURE!

File: cpp_algorithms/trivial_structs/queue_deque.cpp
#include <iostream>
#include <deque>
#include <queue>

using namespace std;

// Deque
int main_deque() {
    deque<int> dq;

    // append()
    dq.push_back(10);

    // appendleft()
    dq.push_front(20);

    // peek_left()
    cout << dq.front() << '\n';

    // peek_right()
    cout << dq.back() << '\n';

    // pop()
    dq.pop_back();

    // popleft()
    dq.pop_front();

    return 0;
}

int main_queue() {
    queue<int> q;

    // enqueue
    q.push(10);
    q.push(20);
    q.push(30);

    // peek
    cout << q.front() << '\n';   // 10

    // dequeue
    q.pop();

    cout << q.front() << '\n';   // 20

    return 0;
}

// Yeah python, that easy.
